doc/develop/api/api_conventions.rst

```diff
@@ -0,0 +1,100 @@
+Zephyr's API space includes the following areas:
+
+- C programming interface (function prototypes, structures, and macros), provided in C header files
+- Configuration system (Kconfig)
+- Hardware description system (Devicetree)
+
+Other areas, such as the build system (CMake), may also be seen as Zephyr's API. However, any area
+that is not explicitly listed herein is not considered Zephyr's API as of today.
+
+API classification
+******************
+The files that define the Zephyr APIs may contain symbols intended for different usage. The intended
+API usage is designated by the API class. Zephyr APIs are classified as:
+
+Private
+    These APIs are intended for use within boundary of a :term:`software component`. Private APIs
+    defined in Zephyr tree are not subject to :ref:`api_lifecycle`. Therefore, can be changed or
+    removed at any time. Changes to the private API may not be documented.
+
+Internal
+    In general, these APIs are intended for use only between certain software components that are
+    located in Zephyr tree. The context where the API is called or implemented is well defined. For
+    example, functions prefixed with `arch_` are intended for use by the Zephyr kernel to invoke
+    architecture specific code. An API is classified as an Internal on a case by case basis.
+    Internal APIs should not be used by out-of-tree code. Internal APIs are not subject
+    to :ref:`api_lifecycle`. Therefore, can be changed or removed at any time. Changes to the
+    internal APIs must be documented in the migration guide.
+
+Public
+    These APIs intended for use from any :term:`software component`. Public APIs may be used in-tree
+    and out-of-tree. Public APIs are subject to the :ref:`api_lifecycle`. Therefore, changes to an
+    API are introduced and documented according to the rules defined for the API life state.
+
+Note, that only APIs defined in Zephyr tree are subject to :ref:`api_lifecycle`. External projects
+used as :ref:`modules` may define their own rules for API lifecycle.
+
+Zephyr is a constantly evolving project and API classification may change over time. A Private or
+Internal API may be promoted to Internal or Public API, respectively. Zephyr users are encouraged to
+follow :ref:`rfcs` process to recommend changes in API classification.
+
+The following sections provide guidelines on how to identify the class of an API depending on its
+type.
+
+Header files contents
+=====================
+Private
+    Functions and data types declared in header files located in
+    :zephyr_file:`include/zephyr/private/`. In addition, private symbols are prefixed with ``z_``.
+    Due to historical reasons some APIs prefixed with ``z_`` are public.
+
+Internal
+    Functions and data types declared :zephyr_file:`include/zephyr/internal`. In addition, Internal
+    APIs must use ``@internal`` doxygen command.
+
+Public
+    Functions and data types declared in header files located in :zephyr_file:`include/zephyr/`.
```

Comment:
There was a proposal from @henrikbrixandersen or @nashif to use Doxygen as the ultimate source of truth instead of the location of the header. I personally prefer using the location, but I do agree that it is non-trivial to do so, because sometimes you need different API types to coexist in the same folder.
If we go with Doxygen classification, then this whole section should change to either remove paths altogether, or else at least highlight that the criteria is Doxygen-based and not path-based